# ANMI020 Spatial microdiversity

```{r load_libraries, comment="", message=FALSE, warning=FALSE}
library(tidyverse)
library(vegan)
library(adespatial)
library(ade4)
library(zCompositions)
library(ape)
library(spdep)
library(lmPerm)
library(tinytable)
library(ggh4x)
select <- dplyr::select

```

```{r load_data, comment="", message=FALSE, warning=FALSE}
ani_dist <- read_tsv("data/GPB_bin_000118_pairwise_ani.tsv") %>% 
  column_to_rownames(var="SampleID") %>% 
  as.dist()

attr(ani_dist, "Labels") <- read_tsv("data/id_to_sample.tsv") %>%
  arrange(ID) %>%
  pull(sample)

metadata <- read_tsv("data/alpha_div_filtered.tsv") %>% 
  filter(
    batch %in% c("MSEB0011"),
    type_binomial %in% c("positive"),
    genome_count_filtering30 %in% c("retained"),
    !is.na(Xcoord))

ani_dist_pairwise <- ani_dist %>%
  as.matrix() %>% 
  as.data.frame() %>% 
  .[metadata$microsample, metadata$microsample] %>%
  rownames_to_column("Sample1") %>%
  pivot_longer(-Sample1, names_to = "Sample2", values_to = "Distance") %>%
  filter(! Sample1 %in% c("M301055","M301090","M300840")) %>% 
  filter(! Sample2 %in% c("M301055","M301090","M300840"))
```

```{r heatmap_all, comment="", message=FALSE, warning=FALSE}
ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  mutate(sample.x=substr(cryosection.x, 1, 4)) %>% 
  mutate(sample.y=substr(cryosection.y, 1, 4)) %>% 
  mutate(cryosection.x=factor(cryosection.x,levels=c("G103bI301B","G103bI301A","G121eI103A","G121eI103B"))) %>% 
  mutate(sample.x=factor(sample.x,levels=c("G103","G121"))) %>% 
  mutate(cryosection.y=factor(cryosection.y,levels=c("G121eI103B","G121eI103A","G103bI301A","G103bI301B"))) %>% 
  mutate(sample.y=factor(sample.y,levels=c("G121","G103"))) %>% 
  ggplot(aes(x = Sample1, y = Sample2, fill = Distance)) +
  geom_tile() +
  scale_fill_viridis_c(limits = c(0.98, 1)) +
  facet_nested(sample.y + cryosection.y ~ sample.x + cryosection.x, scale="free")+
  #theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(title = "Distance Matrix Heatmap", x = "Sample", y = "Sample")
```

```{r ordination_all, comment="", message=FALSE, warning=FALSE}
nmds <- ani_dist_pairwise %>%
  pivot_wider(names_from = Sample2, values_from = Distance) %>%
  column_to_rownames("Sample1") %>%
  as.matrix() %>%
  as.dist() %>% 
  vegan::metaMDS(., trymax = 500, k = 2, trace=0) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "microsample") %>% 
  left_join(metadata %>% select(microsample,cryosection), by="microsample") %>% 
  mutate(animal=substr(cryosection, 1, 4)) %>% 
  filter(microsample != "M300804")

nmds %>% 
ggplot(aes(x = NMDS1, y = NMDS2, color=animal, shape=cryosection)) +
  geom_point() +
  theme_minimal()
```

## G121eI103A

```{r mantel_test_G121eI103A, comment="", message=FALSE, warning=FALSE}
ani_distance <- ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  filter(cryosection.x =="G121eI103A") %>% 
  filter(cryosection.y =="G121eI103A") %>%
  mutate(Distance=1-Distance) %>%
  select(Sample1,Sample2,Distance) %>% 
  pivot_wider(names_from = Sample2, values_from = Distance) %>%
  column_to_rownames("Sample1") %>%
  as.matrix() %>%
  as.dist()
  
spatial_distance <- metadata %>% 
    filter(cryosection =="G121eI103A") %>% 
    select(microsample,Xcoord,Ycoord) %>% 
    column_to_rownames("microsample") %>%
    as.matrix() %>%
    dist(method = "euclidean")

mantel(ani_distance, spatial_distance, method = "pearson", permutations = 999)

```

```{r distance_decay_G121eI103A, comment="", message=FALSE, warning=FALSE}
ani_distance <- ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  filter(cryosection.x =="G121eI103A") %>% 
  filter(cryosection.y =="G121eI103A") %>%
  mutate(Distance=1-Distance) %>%
  select(Sample1,Sample2,Distance) %>% 
  pivot_wider(names_from = Sample2, values_from = Distance) %>%
  column_to_rownames("Sample1") %>%
  as.matrix() %>%
  as.dist()
  
spatial_distance <- metadata %>% 
    filter(cryosection =="G121eI103A") %>% 
    select(microsample,Xcoord,Ycoord) %>% 
    column_to_rownames("microsample") %>%
    as.matrix() %>%
    dist(method = "euclidean")

ani_distance_G121eI103A <-ani_distance %>% 
  as.matrix() %>% as.data.frame() %>% 
  rownames_to_column("Sample1") %>%
  pivot_longer(-Sample1, names_to = "Sample2", values_to = "ANI")

spatial_distance_G121eI103A <- spatial_distance %>% 
  as.matrix() %>% as.data.frame() %>% 
  rownames_to_column("Sample1") %>%
  pivot_longer(-Sample1, names_to = "Sample2", values_to = "Distance")

ani_distance_G121eI103A %>% 
  left_join(spatial_distance_G121eI103A, by=join_by("Sample1"=="Sample1","Sample2"=="Sample2")) %>% 
  ggplot(aes(x = Distance, y = ANI)) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "lm", se = TRUE, color = "blue") +
  theme_minimal() +
  labs(
    title = "Distance-Decay Relationship",
    x = "Spatial Distance",
    y = "Genetic Distance (1 - ANI)"
  )

ani_distance_G121eI103A %>% 
  left_join(spatial_distance_G121eI103A, by=join_by("Sample1"=="Sample1","Sample2"=="Sample2")) %>% 
  lm(ANI ~ Distance, data = .) %>% 
  summary()
```

```{r distance_ani_relationship, comment="", message=FALSE, warning=FALSE}
ani_links <- ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  filter(cryosection.x =="G121eI103A") %>% 
  filter(cryosection.y =="G121eI103A") %>%
  mutate(Distance=1-Distance) %>%
  select(Sample1,Sample2,Distance) %>%
  left_join(metadata %>% select(microsample,Xcoord,Ycoord), by = c("Sample1" = "microsample")) %>%
  rename(X1 = Xcoord, Y1 = Ycoord) %>%
  left_join(metadata %>% select(microsample,Xcoord,Ycoord), by = c("Sample2" = "microsample")) %>%
  rename(X2 = Xcoord, Y2 = Ycoord)

ggplot() +
  geom_segment(data = ani_links, aes(x = X1, y = Y1, xend = X2, yend = Y2, color = Distance), alpha = 0.6) +
  geom_point(data = metadata %>% 
                  filter(cryosection =="G121eI103A") %>% 
                  filter(microsample != "M300840") %>% 
                  select(microsample,Xcoord,Ycoord)
             , aes(x = Xcoord, y = Ycoord), size = 3, color = "black") +
  scale_color_viridis_c(name = "ANI Distance", option = "inferno", limits = c(0, 0.01), direction = -1) +
  theme_minimal() +
  coord_fixed() +
  labs(
    title = "Spatial Network Colored by ANI Distance",
    x = "X coordinate", y = "Y coordinate"
  )
```

## G103bI301A

```{r mantel_test_G103bI301A, comment="", message=FALSE, warning=FALSE}
ani_distance <- ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  filter(cryosection.x =="G103bI301A") %>% 
  filter(cryosection.y =="G103bI301A") %>%
  mutate(Distance=1-Distance) %>%
  select(Sample1,Sample2,Distance) %>% 
  pivot_wider(names_from = Sample2, values_from = Distance) %>%
  column_to_rownames("Sample1") %>%
  as.matrix() %>%
  as.dist()
  
spatial_distance <- metadata %>% 
    filter(cryosection =="G103bI301A") %>% 
    select(microsample,Xcoord,Ycoord) %>% 
    column_to_rownames("microsample") %>%
    as.matrix() %>%
    dist(method = "euclidean")

mantel(ani_distance, spatial_distance, method = "pearson", permutations = 999)

```

```{r distance_decay_G121eI103A, comment="", message=FALSE, warning=FALSE}
ani_distance <- ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  filter(!Sample1 %in% c("M301068","M301084","M301085")) %>% 
  filter(!Sample2 %in% c("M301068","M301084","M301085")) %>% 
  filter(cryosection.x =="G103bI301A") %>% 
  filter(cryosection.y =="G103bI301A") %>%
  mutate(Distance=1-Distance) %>%
  select(Sample1,Sample2,Distance) %>% 
  pivot_wider(names_from = Sample2, values_from = Distance) %>%
  column_to_rownames("Sample1") %>%
  as.matrix() %>%
  as.dist()
  
spatial_distance <- metadata %>% 
    filter(cryosection =="G103bI301A") %>% 
    select(microsample,Xcoord,Ycoord) %>% 
    column_to_rownames("microsample") %>%
    as.matrix() %>%
    dist(method = "euclidean")

ani_distance_G121eI103A <-ani_distance %>% 
  as.matrix() %>% as.data.frame() %>% 
  rownames_to_column("Sample1") %>%
  pivot_longer(-Sample1, names_to = "Sample2", values_to = "ANI")

spatial_distance_G121eI103A <- spatial_distance %>% 
  as.matrix() %>% as.data.frame() %>% 
  rownames_to_column("Sample1") %>%
  pivot_longer(-Sample1, names_to = "Sample2", values_to = "Distance")

ani_distance_G121eI103A %>% 
  left_join(spatial_distance_G121eI103A, by=join_by("Sample1"=="Sample1","Sample2"=="Sample2")) %>% 
  ggplot(aes(x = Distance, y = ANI)) +
  geom_point(alpha = 0.5) +
  geom_smooth(method = "lm", se = TRUE, color = "blue") +
  theme_minimal() +
  labs(
    title = "Distance-Decay Relationship",
    x = "Spatial Distance",
    y = "Genetic Distance (1 - ANI)"
  )

ani_lm <- ani_distance_G121eI103A %>% 
  left_join(spatial_distance_G121eI103A, by=join_by("Sample1"=="Sample1","Sample2"=="Sample2")) %>% 
  lm(ANI ~ Distance, data = .) 

ani_lm %>% 
  summary()

# Predicted

proximal <- predict(ani_lm, newdata = tibble(Distance = c(1:500))) 
proximal %>% mean()
proximal %>% sd()

distal <- predict(ani_lm, newdata = tibble(Distance = c(3501:4000)))
distal %>% mean()
distal %>% sd()

bind_rows(tibble(ani=proximal,distance="proximal"),tibble(ani=distal,distance="distal")) %>% 
  ggplot(aes(x=distance, y=ani, group=distance)) + 
    geom_boxplot()


# Observed
proximal <- ani_distance_G121eI103A %>% 
  left_join(spatial_distance_G121eI103A, by=join_by("Sample1"=="Sample1","Sample2"=="Sample2")) %>% 
  filter(Distance < 500) %>% 
  mutate(distance="proximal") %>% 
  select(ANI,distance)

distal <- ani_distance_G121eI103A %>% 
  left_join(spatial_distance_G121eI103A, by=join_by("Sample1"=="Sample1","Sample2"=="Sample2")) %>% 
  filter(Distance > 3000) %>% 
  mutate(distance="distal") %>% 
  select(ANI,distance)

bind_rows(proximal,distal) %>% 
  group_by(distance) %>% 
  summarise(mean(ANI))

bind_rows(proximal,distal) %>% 
  ggplot(aes(x=distance, y=ANI, group=distance)) + 
    geom_boxplot()
```

```{r distance_ani_relationship, comment="", message=FALSE, warning=FALSE}
ani_links <- ani_dist_pairwise %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample1"=="microsample")) %>% 
  left_join(metadata %>% select(microsample,cryosection),by=join_by("Sample2"=="microsample")) %>% 
  filter(!Sample1 %in% c("M301068","M301084","M301085")) %>% 
  filter(!Sample2 %in% c("M301068","M301084","M301085")) %>% 
  filter(cryosection.x =="G103bI301A") %>% 
  filter(cryosection.y =="G103bI301A") %>%
  mutate(Distance=1-Distance) %>%
  select(Sample1,Sample2,Distance) %>%
  left_join(metadata %>% select(microsample,Xcoord,Ycoord), by = c("Sample1" = "microsample")) %>%
  rename(X1 = Xcoord, Y1 = Ycoord) %>%
  left_join(metadata %>% select(microsample,Xcoord,Ycoord), by = c("Sample2" = "microsample")) %>%
  rename(X2 = Xcoord, Y2 = Ycoord)

ggplot() +
  geom_segment(data = ani_links, aes(x = X1, y = Y1, xend = X2, yend = Y2, color = Distance), alpha = 0.6) +
  geom_point(data = metadata %>% 
                  filter(cryosection =="G103bI301A") %>% 
                  filter(! microsample %in% c("M301068","M301084","M301085")) %>% 
                  select(microsample,Xcoord,Ycoord)
             , aes(x = Xcoord, y = Ycoord), size = 3, color = "black") +
  scale_color_viridis_c(name = "ANI Distance", option = "inferno", limits = c(0, 0.005), direction = -1) +
  theme_minimal() +
  coord_fixed() +
  labs(
    title = "Spatial Network Colored by ANI Distance",
    x = "X coordinate", y = "Y coordinate"
  )
```

```{r vcf, comment="", message=FALSE, warning=FALSE}
vcf <- read_tsv("/Users/anttonalberdi/Downloads/GPB_bin_000118.vcf") %>% 
  select(CHROM, POS) %>% 
  mutate(value=1)

annotations <- read_tsv("/Users/anttonalberdi/Downloads/dram.annotations.tsv.gz") %>% 
  filter(fasta == "GPB:bin_000118")

SNP_bins <- vcf %>% 
  mutate(pos_bin = floor((POS - 1) / 1000) * 1000 + 1) %>%  # create bins like 1, 1001, 2001...
  group_by(pos_bin) %>%
  summarise(total_value = sum(value), .groups = 'drop') %>%
  arrange(pos_bin) %>% 
  arrange(-total_value)
  
SNP_bins %>% 
  ggplot(aes(x=pos_bin,y=total_value)) +
    geom_point()

# Step 1: Add window end position
SNP_bins2 <- SNP_bins %>%
  mutate(window_start = pos_bin,
         window_end = pos_bin + 1000 - 1)

overlapping_genes <- SNP_bins2 %>%
  rowwise() %>%
  mutate(gene_info = list(
    annotations %>%
      filter(start_position <= window_end, end_position >= window_start)
  )) %>%
  unnest(gene_info) %>% 
  select(pos_bin,total_value,ko_id)

overlapping_genes %>% print(n=30)

overlapping_genes %>% 
  group_by(ko_id) %>% 
  summarise(count=n()) %>% 
  arrange(-count)
```